enhancements/single-node-developer-cluster-profile.md

```diff
@@ -0,0 +1,514 @@
+  specify the cluster profile, to the standard OpenShift installer
+  currently used to build CRC installation images. Evolution of the
+  existing installer, SNC tooling and other CRC components is
+  expected.
+* Functional changes for OpenShift. This enhancement proposal does not
+  result in any functional changes to clusters that do not utilize the
+  single-node-developer cluster profile.
+
+## Proposal
+
+The single-node-developer cluster profile will be implemented in two
+phases. The first phase introduces the new cluster profile, and
+enables some simplification of the SNC image build process by
+utilizing the profile to control which components are included in the
+cluster. The second phase focuses on optimizing the resource
+utilization of the components included in the cluster and may further
+simplify the SNC image build process.
+
+### Phase One
+
+In the initial phase, the following annotation will be added to the
+manifest files for all components that are to be included in single
+node developer clusters:
+
+```
+include.release.openshift.io/single-node-developer=true
+```
+
+This annotation will need to be added to the manifests for all the
+OpenShift components that are not currently removed by the snc.sh
+script during the CRC build process. The components that are currently
+removed, and thus should not need the annotation added, include:
+
+* most operators running in the openshift-monitoring namespace
+* most operators running in the openshift-insights namespace
+* machine-api-operator
+* Machine-config-operator (may need to run once, requiring the annotation)
+* openshift-cloud-credential-operator
+* openshift-cluster-storage-operator
+
+Additional components that are not currently removed by the SNC
+tooling might be also be determined to be unnecessary, and thus not
+require the annotation. Possibilities include:
+
+* openshift-cluster-node-tuning-operator
+* openshift-multus
+
+Once PRs adding this annotation have been merged to all the necessary
+component repositories, the snc.sh script will be updated to tell the
+OpenShift installer to use the single-node-developer cluster profile,
+and to remove the code that deletes the components that are not
+annotated. At the conclusion of this initial phase, the somewhat
+simplified CRC build tooling should produce a cluster equivalent to
+that currently produced, and all existing CRC tests should pass. Phase
+one is expected to be completed during development of the 4.7
+OpenShift release.
+
+From this point forward, any new components being added to OpenShift
+will not be included in CRC unless a manifest with the
+single-node-developer annotation is included in the new component. The
+CRC team will need to engage with other OpenShift teams to help decide
+whether or not new components should be included in the profile, and
+if so, how they should be configured.
+
+### Phase Two
+
+In the second phase, various OpenShift components that are included in
```

Comment:
Not all cluster operators allow their behavior to be modified declaratively - some behavior is implemented as code. I think this implies that the desired behavior (e.g. reduced resource usage, non-HA) will require more than just new manifests, and that the testing burden involved is potentially non-trivial.

I'm also not sure it makes sense to use a feature-specific annotation as described (i.e. `single-node-developer`true`) given proposals like #504. Might it make more sense to agree on one or more feature-agnostic markers (e.g. `resource-constrained-cluster`) where the resulting outcome could be agreed upon? This would seem to simplify life for component makers by allowing them to implement support for shared requirements rather than having to worry about every feature independently.

Reply:
> Not all cluster operators allow their behavior to be modified declaratively - some behavior is implemented as code. I think this implies that the desired behavior (e.g. reduced resource usage, non-HA) may require more than just new manifests, and that the testing burden involved is potentially non-trivial.

Yes, we are aware that during the second phase, code changes are likely to be needed in some components. However, once the code is in place, the appropriate code path can be selected through a manifest (for example by adding a commandline option for it, or the operator could directly read what is the active profile).